src/util/checkpt/fd_checkpt.c

```diff
@@ -0,0 +1,911 @@
+        FD_LOG_WARNING(( "mmio_sz too small" ));
+        checkpt->frame_style = -1; /* failed */
+        return FD_CHECKPT_ERR_IO;
+      }
+
+      memcpy( checkpt->mmio.mem + off, buf, sz );
+      off += sz;
+
+    } else { /* streaming mode */
+
+      ulong wbuf_used = checkpt->wbuf.used;
+
+      ulong wsz_max = wbuf_used + sz;
+      if( FD_UNLIKELY( wsz_max<sz ) ) {
+        FD_LOG_WARNING(( "sz overflow" ));
+        checkpt->frame_style = -1; /* failed */
+        return FD_CHECKPT_ERR_IO;
+      }
+
+      int err = fd_io_buffered_write( checkpt->fd, buf, sz, checkpt->wbuf.mem, checkpt->wbuf.sz, &wbuf_used );
+      if( FD_UNLIKELY( err ) ) {
+        FD_LOG_WARNING(( "fd_io_buffered_write failed (%i-%s)", err, fd_io_strerror( err ) ));
+        checkpt->frame_style = -1; /* failed */
+        return FD_CHECKPT_ERR_IO;
+      }
+
+      if( FD_UNLIKELY( wsz_max<wbuf_used ) ) {
+        FD_LOG_WARNING(( "unexpected buffered write size" ));
+        checkpt->frame_style = -1; /* failed */
+        return FD_CHECKPT_ERR_IO;
+      }
+
+      ulong wsz = wsz_max - wbuf_used;
+
+      off += wsz;
+      if( FD_UNLIKELY( off<wsz ) ) {
+        FD_LOG_WARNING(( "checkpt sz overflow" ));
+        checkpt->frame_style = -1; /* failed */
+        return FD_CHECKPT_ERR_IO;
+      }
+
+      checkpt->wbuf.used = wbuf_used;
+
+    }
+
+    break;
+  }
+
+# if FD_HAS_LZ4
+  case FD_CHECKPT_FRAME_STYLE_LZ4: {
+    LZ4_stream_t * lz4 = (LZ4_stream_t *)checkpt->lz4;
+
+    if( fd_checkpt_private_is_mmio( checkpt ) ) { /* mmio mode, app dependent branch prob */
+      char * mmio      = checkpt->mmio.mem;
+      ulong  mmio_sz   = checkpt->mmio.sz;
+
+      char const * chunk = (char const *)buf;
+      do {
+        ulong mmio_rem = mmio_sz - off;
+        if( FD_UNLIKELY( !mmio_rem ) ) {
+          FD_LOG_WARNING(( "mmio_sz too small" ));
+          checkpt->frame_style = -1; /* failed */
+          return FD_CHECKPT_ERR_IO;
+        }
+
+        ulong chunk_usz     = fd_ulong_min( sz, FD_CHECKPT_PRIVATE_CHUNK_USZ_MAX );
+        ulong chunk_csz_max = fd_ulong_min( FD_CHECKPT_PRIVATE_CSZ_MAX( chunk_usz ), mmio_rem );
+
+        /* At this point, chunk_usz is in [1,CHUNK_USZ_MAX] and
+           chunk_csz_max is in [1,CSZ_MAX( USZ_MAX )].  chunk_csz_max is
+           also at most the amount of space remaining in the mmio
+           region.
+
+           FIXME: DOES CHUNK_CSZ_MAX (DSTCAPACITY) AFFECT HOW LZ4 DOES
+           ITS COMPRESSION/DECOMPRESSION?  THE APIS SUGGEST NO AND THIS
+           ASSUMES IT DOESNT.  IF IT DOES, THIS CALL AND VARIOUS CALLS
+           BELOW MIGHT NEED TO BE ADJUST TO PROVIDE IDENTICAL VALUES FOR
+           DST_CAPACITY BETWEEN STREAMING AND MMIO AND BETWEEN CHECKPT
+           AND RESTORE. */
+
+        /* Note: 1 is default LZ4 acceleration */
+        int _chunk_csz = LZ4_compress_fast_continue( lz4, chunk, mmio + off, (int)chunk_usz, (int)chunk_csz_max, 1 );
```

Comment:
This is already known, but stating just for the record - The `LZ4_compress_fast_continue` is not safe with untrusted input